### ISSUE TITLE
Deadlock in `signalReceivers.Stop`

### ISSUE THREAD

**Issue by ogaca-dd** [uber-go/fx#1219]
**Describe the bug**
There is a deadlock in [signalReceivers.Stop](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L120):
- `syscall.SIGINT` is sent to the process
- `Shutdowner.Shutdown()` is called to stop the application 
- [signalReceivers.Stop](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L120) is called and takes the lock [m](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L121)
- [signalReceivers.Stop](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L120) sends a shutdown request ([recv.shutdown <- struct{}{}](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L130)) and blocks on the [select](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L132)
- In [signalReceivers.relayer](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L84), the branch [case signal := <-recv.signals:](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L92) is executed before [case <-recv.shutdown](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L90)
- The branch [case signal := <-recv.signals:](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L92) calls [signalReceivers.Broadcast](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L177)
- [signalReceivers.Broadcast](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L177) takes the same lock [`m`](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L178) as [signalReceivers.Stop](https://github.com/uber-go/fx/blob/v1.22.0/signal.go#L121)

There is a deadlock.

**To Reproduce**
Run the following code:
```go
func main() {
	fx.New(fx.Invoke(func(shutdowner fx.Shutdowner) {
		go func() {
			_ = syscall.Kill(syscall.Getpid(), syscall.SIGINT)
			_ = shutdowner.Shutdown()
		}()
	})).Run()
}
```

**Expected behavior**
No deadlock

**Comment by JacobOaks** [comment]
Hey @ogaca-dd thanks for the report, let me look into this.

Can I ask what kinds of situations you're experiencing where a shutdown signal and an OS signal are being sent at roughly the same time?

**Comment by JacobOaks** [comment]
Hey @ogaca-dd - I understand. My question is why is your application calling `shutdowner.Shutdown()` after it gets killed? The example you provide is pretty contrived so I'm just trying to better understand the impact here.

**Comment by ogaca-dd** [comment]
The following code is the simplest code I found to reproduce the issue.
```go 
func main() {
	fx.New(fx.Invoke(func(shutdowner fx.Shutdowner) {
		go func() {
			_ = syscall.Kill(syscall.Getpid(), syscall.SIGINT)
			_ = shutdowner.Shutdown()
		}()
	})).Run()
}
```

Our real code looks like 
```go
        sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGPIPE)
	for signo := range sigChan {
		switch signo {
		case syscall.SIGINT, syscall.SIGTERM:
			log.Infof("Received signal %d (%v)", signo, signo)
			 shutdowner.Shutdown()
			return
                }
               // Some code here
       }
}
```
I know that fx already handles signals and this code shouldn't exist (This code was not removed during our migration to fx).

Anyway, the code I pointed contains a dead lock which was tricky to find. In our case, this freeze happens randomly:
- The frequency of the issue depends on the OS 
- Adding debugging logs decreases significantly the occurence of the issue which make it harder to catch.

I understand it may not be a correct usage of fx but as a user I don't expect to have my application freeze randomly in a such case.

**Comment by JacobOaks** [comment]
Gotcha. Yeah I was able to reproduce this locally and I will try to figure out the best way to update the signal handlers to prevent this, but yes - as you mentioned - using `app.Run` will cause Fx to respond to signals automatically. So if you want to workaround this for now you can either:
* Not manually call shutdown when you receive a signal
* Use `app.Start` and then call `app.Stop` when a signal comes into `sigChan`. Using Start/Stop instead of Run in this way will no longer cause Fx to register its own signal handlers as of v1.22.1.

**Comment by ogaca-dd** [comment]
Thanks for your quick reply and for the workarounds.

**Comment by abhinav** [comment]
@JacobOaks is this resolved by #1220?

**Comment by JacobOaks** [comment]
Fixed by #1220.
